arch/x86_64/core/x86_64.c

```diff
@@ -0,0 +1,193 @@
+	cpu_init[cpu_num].esp = (int)(long)(sz + (char *)stack);
+
+	/* This is our flag to the spinning CPU.  Do this last */
+	cpu_init[cpu_num].fn = fn;
+}
+
+#ifdef CONFIG_IRQ_OFFLOAD
+static irq_offload_routine_t offload_fn;
+static void *offload_arg;
+
+static void irq_offload_handler(void *arg, int err)
+{
+	ARG_UNUSED(arg);
+	ARG_UNUSED(err);
+	offload_fn(offload_arg);
+}
+
+void irq_offload(irq_offload_routine_t fn, void *arg)
+{
+	offload_fn = fn;
+	offload_arg = arg;
+
+	_apic.ICR_HI = (struct apic_icr_hi) {};
+	_apic.ICR_LO = (struct apic_icr_lo) {
+		.delivery_mode = FIXED,
+		.vector = CONFIG_IRQ_OFFLOAD_VECTOR,
+		.shorthand = SELF,
+	};
+}
+#endif
+
+/* Default.  Can be overridden at link time by a timer driver */
+void __weak x86_apic_timer_isr(void *arg, int code)
+{
+	ARG_UNUSED(arg);
+	ARG_UNUSED(code);
+}
+
+/* Called from xuk layer on actual CPU start */
+void _cpu_start(int cpu)
+{
+	xuk_set_f_ptr(cpu, &_kernel.cpus[cpu]);
+
+	/* Set up the timer ISR, but ensure the timer is disabled */
+	xuk_set_isr(INT_APIC_LVT_TIMER, 13, x86_apic_timer_isr, 0);
+	_apic.INIT_COUNT = 0;
+
+#ifdef CONFIG_IRQ_OFFLOAD
+	xuk_set_isr(XUK_INT_RAW_VECTOR(CONFIG_IRQ_OFFLOAD_VECTOR),
+		    -1, irq_offload_handler, 0);
+#endif
+
+	if (cpu <= 0) {
+		/* The SMP CPU startup function pointers act as init
+		 * flags.  Zero them here because this code is running
+		 * BEFORE .bss is zeroed!  Should probably move that
+		 * out of _Cstart() for this architecture...
+		 */
+		for (int i = 0; i < CONFIG_MP_NUM_CPUS; i++) {
+			cpu_init[i].fn = 0;
+		}
+
+		/* Enter Zephyr */
+		_Cstart();
+
+	} else if (cpu < CONFIG_MP_NUM_CPUS) {
+		/* SMP initialization.  First spin, waiting for
+		 * _arch_start_cpu() to be called from the main CPU
+		 */
+		while (!cpu_init[cpu].fn) {
+		}
+
+		/* Enter Zephyr, which will switch away and never return */
+		cpu_init[cpu].fn(0, cpu_init[cpu].arg);
+	}
+
+	/* Spin forever as a fallback */
+	while (1) {
+	}
+}
+
+/* Returns the initial stack to use for CPU startup on auxiliary (not
+ * cpu 0) processors to the xuk layer, which gets selected by the
+ * non-arch Zephyr kernel and stashed by _arch_start_cpu()
+ */
+unsigned int _init_cpu_stack(int cpu)
+{
+	return cpu_init[cpu].esp;
+}
+
+int _arch_irq_connect_dynamic(unsigned int irq, unsigned int priority,
+			      void (*routine)(void *parameter), void *parameter,
+			      u32_t flags)
+{
+	ARG_UNUSED(flags);
+	__ASSERT(priority >= 2 && priority <= 15,
+		 "APIC interrupt priority must be 2-15");
+
+	xuk_set_isr(irq, priority, (void *)routine, parameter);
+	return 0;
+}
+
+void _arch_irq_disable(unsigned int irq)
+{
+	xuk_set_isr_mask(irq, 1);
+}
+
+void _arch_irq_enable(unsigned int irq)
+{
+	xuk_set_isr_mask(irq, 0);
+}
+
+void x86_apic_set_timeout(u32_t cyc_from_now)
+{
+	_apic.INIT_COUNT = cyc_from_now;
+}
+
+const NANO_ESF _default_esf;
+
+void _NanoFatalErrorHandler(unsigned int reason, const NANO_ESF *esf)
+{
+	printk("Fatal Error reason %d esf %p\n", reason, esf);
```

Comment:
Will fix.  Can't really adapt anything from x86 because the format of the stack frame is entirely different.  And the caller-save registers are going to be problematic to do robustly (the ISR entry doesn't push them until after it has invoked the handler, and even then only if it's going to be restoring to a different thread and needs to pickle them for later), but I'll figure something out.

Reply:
>  And the caller-save registers are going to be problematic to do robustly (the ISR entry doesn't push them until after it has invoked the handler

IIRC in x86's excstub.S we also push the caller-save registers so that the NANO_ESF has everything of interest in it.